### PR TITLE
Introduce 3.0 undefine queries

### PR DESCRIPTION
**Pull request by farost**
## Release notes: product changes
We introduce `undefine` queries with the new "targeted" syntax to enable safe and transparent schema concept undefinitions. 

If you want to undefine the whole type, you can just say:
```
undefine 
  person;
```

If you want to undefine a capability from **somewhere**, use `undefine **capability** from **somewhere**`. 
It consistently works with `owns` (`person` is preserved, only the `owns` is undefined):
```
undefine
  owns name from person;
```
annotations of `owns` (`person owns name` is preserved, only `@regex(...)` (you don't need to specify the regex's argument) is undefined):
```
undefine
  @regex from person owns name;
```
and any other capability, even specialisation:
```
undefine
  as parent from fathership relates father;
```

Want to undefine multiple concepts in one go? Worry not!
```
undefine
  person;
  relates employee from employment;
  @regex from name;
  @values from email;
```

The error messages in all definition queries (`define`, `redefine`, and `undefine`) were enhanced, and the respective `query/language` BDD tests were introduced in the CI with `concept` unit tests.

Additional fixes: 
* answers of `match` capability queries like `match $x owns $y`, `match $x relates $y`, and `$match $x plays $y` now include transitive capabilities;
* `define` lets you write multiple declarations of an undefined type, specifying its kind **anywhere**, even in the last mention of this type;
* failures in schema queries for schema transactions no longer lead to freezes of the following opened transactions;
* no more crashes on long string attributes deletion with existing `has` edges;

## Implementation

The `undefine` feature implementation follows the inherited architecture and logic of `define` and `redefine` queries: 
* traverse the parsed undefinables (in reversed order to `define` to start from "leaves") 
* for types, check their definition status (doesn't exist at all / exists as requested / exists different (different sub, different ordering) -> return an error or do undefine
* for capabilities of types, find types and error if not found -> check capabilities definition status -> return an error or do undefine
* for capabilities of capabilities, find types and error if not found -> find capabilities and error if not found or different -> check capabilities's capabilities definition status -> return an error or do undefine

For additional fixes, additional comments are left on top of the code when needed.

### CODE REVIEW

File: executor/instruction/owns_executor.rs
Comment (farost) [reply executor/instruction/owns_executor.rs:1]: repaste

Comment (farost) [inline executor/instruction/owns_executor.rs:1]: I refactored all capabilities executors to use a tuple of ObjectType and InterfaceType instead of capabilities.

There was an issue that we saved iterators of duplicated, say, Owns, with the same declared owner of this owns for multiple subowner sub owner and subsubowner sub subowner and subowner2 sub owner, etc. This way, we didn't returned inherited, transitive capabilities.

I don't want to change Concept API to return "not stored" constraints: it is consistent in the way of returning only declared things. And it's too much work now with no clear benefit.

We already use tuples for has, so it's not too dirty, and the approach is quite consistent here, too.

File: executor/instruction/relates_executor.rs
Comment (farost) [reply executor/instruction/relates_executor.rs:1]: repaste

Comment (farost) [inline executor/instruction/relates_executor.rs:1]: I've made relates always use transitive relation types, because we never hide it in 3.0.

The way to query for non-specialised relates should contain @abstract annotation for relates as it's a sign of specialisation.

However, maybe we need to return only the root relation type (the relation type where this relates is declared) somewhere, as it was previously implemented, and I just couldn't find an example where it worked better than with my fixes.

@dmitrii-ubskii notice: I've put all the existing random failures with `relates` into a Discord thread connected to the message with these prs

File: concept/tests/test_statistics.rs
Comment (farost) [inline concept/tests/test_statistics.rs:300]: There was another snapshot trying to commit, so it should be correct.

File: server/service/transaction_service.rs
Comment (farost) [inline server/service/transaction_service.rs:739]: We need to return the transaction into `self` before exiting this method. Otherwise, the transaction is not closed, and the schema lock is not released....

File: database/database.rs
Comment (dmitrii-ubskii) [reply database/database.rs:92]: I'd expect us to keep track of how much time we spent acquiring the lock and then use the adjusted timeout (minus time already spent) later in `recv`.

File: database/database.rs
Comment (dmitrii-ubskii) [inline database/database.rs:101]: How does the other side of the queue handle the `SendError`s which will pop up if this timeout triggers?

File: database/transaction.rs
Comment (dmitrii-ubskii) [inline database/transaction.rs:354]: Item-like macros should use `{}` rather than `()`.
`{}`, `[]`, and `()` are interchangeable in macro invocations, but the convention is `{}` for item-like (e.g. `thread_local! {}`), `()` for function-like (`print!()`), `[]` for array-like (`vec![]`).

File: dependencies/vaticle/repositories.bzl
Comment (dmitrii-ubskii) [inline dependencies/vaticle/repositories.bzl:48]: I believe both of these are approved now.

File: query/undefine.rs
Comment (dmitrii-ubskii) [inline query/undefine.rs:1152]: `TypeDBError` doesn't require these, and has its own impl of `Display` based on provided description. These should be removed.

File: server/service/transaction_service.rs
Comment (dmitrii-ubskii) [reply server/service/transaction_service.rs:434]: We should not use `Error` in variant names. The fact that it's an error is already communicated by the type, and the variant should clarify which kind of error occurred.

File: concept/tests/test_thing.rs
Comment (farost) [reply concept/tests/test_thing.rs:9]: Yes, and it hasn't been cleaned up (which is a common issue in our code base, btw..). Thank you!

File: concept/tests/test_thing.rs
Comment (farost) [reply concept/tests/test_thing.rs:544]: True, I was also confused! I'll rename it to another collection.

File: database/database.rs
Comment (farost) [reply database/database.rs:92]: Cool, I wanted to do it, but decided to wait for your thoughts. I'll implement it this way, then, now.

Comment (farost) [reply database/database.rs:92]: Done, please review in the next commit!

File: query/undefine.rs
Comment (farost) [reply query/undefine.rs:1152]: Huh, I was sure I had an error when I tried removing it (and I saw that it's already implemented in the macro). Now it works, so it's good. Thanks

File: dependencies/vaticle/repositories.bzl
Comment (farost) [reply dependencies/vaticle/repositories.bzl:48]: They are, but I'm waiting for @flyingsilverfin 's review

File: database/database.rs
Comment (farost) [reply database/database.rs:101]: Good point, I didn't think there would be an error in this case. There was a 
`// TODO: if get disconnect error, skip!`

So I implemented it like
`let _skipped_sync_error = notifier.send(()).ok();`
Check the latest commits, please! I might have gotten the idea wrong.

File: database/transaction.rs
Comment (farost) [reply database/transaction.rs:354]: @flyingsilverfin thoughts?

File: concept/tests/test_statistics.rs
Comment (flyingsilverfin) [reply concept/tests/test_statistics.rs:300]: does it even matter?

Comment (flyingsilverfin) [reply concept/tests/test_statistics.rs:300]: I think you should move the statistics assert check to right after the new `may_synchronise()`, and then you can do the commit that fails, and then check the statistics has _not_ changed.

File: concept/tests/test_thing.rs
Comment (flyingsilverfin) [inline concept/tests/test_thing.rs:544]: looks right now yep

File: concept/tests/test_thing.rs
Comment (flyingsilverfin) [inline concept/tests/test_thing.rs:686]: Correct!

File: concept/tests/test_thing.rs
Comment (flyingsilverfin) [inline concept/tests/test_thing.rs:828]: I'd be happy if this is still all working! :D nice

File: concept/type_/mod.rs
Comment (flyingsilverfin) [reply concept/type_/mod.rs:286]: My only thinking is that this has a memory/performance performance profile from the `get_owns_declared`, since it has to allocate a HashSet each time you call it.
1. Do you need this API?
2. Does it guarantee extra Uniqueness (not getting duplicate AttributeTypes) that is useful to you?

If yes to both - i think it might be worth adding a API for this that doesn't rely on the other `get_owns_declared`, and has to be implemented in each implementation -> can be a new or different cache entry to avoid the allocation...

On the other hand, this is the kind of premature optimisation that led us to some technical debt and probably not worth doing! How do we feel about doing a middle ground of prefixing `compute_xxxx` when it's a Types ConceptAPI that does extra allocations and dropping that prefix when it is not? Does that mess up readability too much?

File: concept/type_/relation_type.rs
Comment (flyingsilverfin) [inline concept/type_/relation_type.rs:321]: Similar point for all these basically

File: database/database.rs
Comment (flyingsilverfin) [inline database/database.rs:144]: I've also started using `expect` to make debugging these panics so much easier! We should keep using them.

File: database/database.rs
Comment (flyingsilverfin) [inline database/database.rs:162]: I'd call this `TRY_LOCK_SLEEP_INTERVAL`

File: query/undefine.rs
Comment (flyingsilverfin) [inline query/undefine.rs:1006]: We should generally not have constant strings that we know exist elsewhere - for `sub` should be from a TypeQL Token, same with `as`, `relates`... etc.

File: database/database.rs
Comment (farost) [reply database/database.rs:162]: Gp!

File: query/undefine.rs
Comment (farost) [reply query/undefine.rs:1006]: It kinda makes sense (ignoring the fact that you were the one who introduced these strings first :wink:), although it makes the error message reading even messier, and we won't probably put whole error messages in our bdd checks. I can refactor it as I agree it's cleaner, but it doesn't feel that good.

File: concept/type_/mod.rs
Comment (farost) [reply concept/type_/mod.rs:286]: I was hesitant about creating these APIs. The facts that inclined me to doing it:
* We already have similar "helper" apis with clones, and we are forced to do these clones because we have `MaybeOwns`. Even if we don't do it, we have to clone on the API user (in this case `executor`) side
* We actually do clones, collects, and other "greedy" (if applicable here)  operations for these types/capabilities multiple times in executors
* Without these apis, Dmitrii had an impression that `get_owns` returns just all owns a type has, with this type as the owner if it's an inherited owns. I didn't want to change anything globally, it would break many things (and I'm not sure if it's needed), but we agreed that it's not that bad if we at least had an api to answer the question "can I get all the attributes this type owns?". That's the answer
* The overall usage looked cleaner this way


Based on this, I'm not sure if we need to push any concept api interfaces optimization now since we don't have clear numbers, and with the limited resources, we should prioritize. It looks consistent enough in the context of other concept API calls now, although I agree with the drawbacks we have now.

We can always add these caches, it's easy, but a bit annoying work. Let's do it later and globally. I'm not against of creating a list of hypotheses to check our numbers against later!

File: query/undefine.rs
Comment (farost) [reply query/undefine.rs]: I have pushed an example of what I can do for `sub`, please check it out!

File: query/undefine.rs
Comment (farost) [inline query/undefine.rs:1015]: @flyingsilverfin Is it better now? Check `{key}`

File: database/transaction.rs
Comment (farost) [reply database/transaction.rs]: @flyingsilverfin ping

File: database/database.rs
Comment (farost) [reply database/database.rs]: Done

File: query/undefine.rs
Comment (flyingsilverfin) [reply query/undefine.rs:1015]: works for me!

File: query/undefine.rs
Comment (flyingsilverfin) [reply query/undefine.rs]: Yeah i know i was rushing, I think there probably is a better strategy overall where we tie the problem back to an original  TypeQL query part. Although that isn't really always possible!

**Review by flyingsilverfin — APPROVED** [review]
Good stuff

File: dependencies/vaticle/repositories.bzl
Comment (farost) [inline dependencies/vaticle/repositories.bzl:43]: TODO: Update before merge

**Pull request merged**
